web-api/src/storage/property_filter.rs

```diff
@@ -0,0 +1,230 @@
+// Copyright 2023 Xayn AG
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, version 3.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+//! This module contains generic code used by the various aspects of property filtering.
+
+use std::collections::HashMap;
+
+use derive_more::{Display, From, IntoIterator};
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+
+use crate::models::DocumentPropertyId;
+
+#[derive(Debug, Display, PartialEq, Error, Serialize)]
+#[non_exhaustive]
+
+pub(crate) enum IncompatibleUpdate {
+    #[display(fmt = "Property {property} is already defined.")]
+    PropertyIsAlreadyIndexed { property: DocumentPropertyId },
+    #[display(
+        fmt = "Only {allowed} indexed properties including publication_date are allowed, got: {count}"
+    )]
+    TooManyProperties { count: usize, allowed: usize },
+}
+
+//Hint: Currently the API and internal definition match so we use the same type.
+#[derive(Debug, Clone, Default, IntoIterator, Serialize, Deserialize)]
+#[serde(transparent)]
+pub(crate) struct IndexedPropertiesSchemaUpdate(
+    #[into_iterator(owned, ref)] IndexedPropertiesSchema,
+);
+
+impl IndexedPropertiesSchemaUpdate {
+    pub(crate) fn len(&self) -> usize {
+        self.0.len()
+    }
+}
+
+//Hint: Currently the API and internal definition match so we use the same type.
+#[derive(Debug, Clone, Default, IntoIterator, From, Serialize, Deserialize)]
+pub(crate) struct IndexedPropertiesSchema {
+    #[into_iterator(owned, ref)]
+    properties: HashMap<DocumentPropertyId, IndexedPropertyDefinition>,
+}
+
+impl IndexedPropertiesSchema {
+    pub(crate) fn len(&self) -> usize {
+        self.properties.len()
+    }
+
+    pub(crate) fn update(
+        &mut self,
+        schema_update: &IndexedPropertiesSchemaUpdate,
```

Comment:
❓ it looks like we need to clone everything from this argument internally anyways, so should we just take an owned argument instead? it also makes it more clear for the caller.